File: sections/40_data_structures_mixed_contest/problems/e_functional_walk_minimum/solution.py
import sys
from array import array

def main():
    data = list(map(int, sys.stdin.buffer.read().split()))
    n, q = data[:2]
    next_vertex = array('i', (x - 1 for x in data[2:2 + n]))
    value = data[2 + n:2 + 2 * n]
    index = 2 + 2 * n
    levels = 61
    up = [next_vertex]
    first_minimum = (value[next_vertex[v]] for v in range(n))
    minimum = [array('q', first_minimum)]
    for level in range(1, levels):
        previous_up = up[level - 1]
        previous_minimum = minimum[level - 1]
        next_up = (previous_up[previous_up[v]] for v in range(n))
        up.append(array('i', next_up))
        next_minimum = (
            min(
                previous_minimum[v],
                previous_minimum[previous_up[v]],
            )
            for v in range(n)
        )
        minimum.append(array('q', next_minimum))
    output = []
    for _ in range(q):
        vertex = data[index] - 1
        steps = data[index + 1]
        index += 2
        answer = value[vertex]
        bit = 0
        while steps:
            if steps & 1:
                answer = min(answer, minimum[bit][vertex])
                vertex = up[bit][vertex]
            steps //= 2
            bit += 1
        output.append(f'{vertex + 1} {answer}')
    sys.stdout.write('\n'.join(output))
if __name__ == '__main__':
    main()

File: sections/40_data_structures_mixed_contest/problems/e_functional_walk_minimum/solution.cpp
#include <bits/stdc++.h>
using namespace std;

using ll = long long;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, q;
    cin >> n >> q;

    constexpr int LEVELS = 61;
    vector<array<int, LEVELS>> up(n);
    vector<array<ll, LEVELS>> minimum(n);

    for (int vertex = 0; vertex < n; ++vertex) {
        cin >> up[vertex][0];
        --up[vertex][0];
    }

    vector<ll> value(n);
    for (ll& current : value)
        cin >> current;

    for (int vertex = 0; vertex < n; ++vertex)
        minimum[vertex][0] = value[up[vertex][0]];

    for (int level = 1; level < LEVELS; ++level) {
        for (int vertex = 0; vertex < n; ++vertex) {
            int middle = up[vertex][level - 1];
            up[vertex][level] = up[middle][level - 1];
            minimum[vertex][level] = min(
                minimum[vertex][level - 1],
                minimum[middle][level - 1]
            );
        }
    }

    while (q--) {
        int vertex;
        unsigned long long steps;
        cin >> vertex >> steps;
        --vertex;

        ll answer = value[vertex];
        for (int level = 0; level < LEVELS; ++level) {
            if ((steps >> level) & 1ULL) {
                answer = min(answer, minimum[vertex][level]);
                vertex = up[vertex][level];
            }
        }

        cout << vertex + 1 << ' ' << answer << '\n';
    }
}
